File: ruby/sum-of-multiples/sum_of_multiples.rb
class SumOfMultiples
  attr_reader :multiples

  def initialize(*multiples)
    @multiples = multiples
  end

  def to(up_to)
    sum = 0

    (1..(up_to - 1)).each do |i|
      sum += i if multiple?(i)
    end

    sum
  end

  def self.to(up_to)
    new(3, 5).to(up_to)
  end

  private

  def multiple?(i)
    multiples.any? do |multiple|
      i % multiple == 0
    end
  end
end

File: go/sum-of-multiples/sum_of_multiples.go
package summultiples

// MultipleSummer returns a function that returns the sum of all multiples
// particular numbers up to but not including that number.
func MultipleSummer(multiples ...int) func(int) int {
	fn := func(limit int) int {
		var sum int

		for i := 1; i < limit; i++ {
			for _, multiple := range multiples {
				if i%multiple == 0 {
					sum += i
					break
				}
			}
		}

		return sum
	}

	return fn
}
